mat64/cholesky.go

```diff
@@ -178,3 +178,162 @@ func (s *SymDense) InverseCholesky(chol *Cholesky) error {
+	//   modifying matrix factorizations. Mathematics of Computation 28(126)
+	//   (1974), Method C3 on page 521
+	//
+	// The implementation is based on LINPACK code
+	// http://www.netlib.org/linpack/dchud.f
+	// http://www.netlib.org/linpack/dchdd.f
+	// and
+	// https://icl.cs.utk.edu/lapack-forum/viewtopic.php?f=2&t=2646
+	//
+	// According to http://icl.cs.utk.edu/lapack-forum/archives/lapack/msg00301.html
+	// LINPACK is released under BSD license.
+	//
+	// See also:
+	// - M. A. Saunders: Large-scale Linear Programming Using the Cholesky
+	//   Factorization. Technical Report Stanford University (1972)
+	//   http://i.stanford.edu/pub/cstr/reports/cs/tr/72/252/CS-TR-72-252.pdf
+	// - Matthias Seeger: Low rank updates for the Cholesky decomposition.
+	//   EPFL Technical Report 161468 (2004)
+	//   http://infoscience.epfl.ch/record/161468
+
+	work := make([]float64, n)
+	blas64.Copy(n, x.RawVector(), blas64.Vector{1, work})
+
+	if alpha > 0 {
+		// Compute rank-1 update.
+		if alpha != 1 {
+			blas64.Scal(n, math.Sqrt(alpha), blas64.Vector{1, work})
+		}
+		umat := c.chol.mat
+		stride := umat.Stride
+		for i := 0; i < n; i++ {
+			// Compute parameters of the Givens matrix that zeroes
+			// the i-th element of x.
+			c, s, r, _ := blas64.Rotg(umat.Data[i*stride+i], work[i])
+			if r < 0 {
+				// Multiply by -1 to have positive diagonal
+				// elemnts.
+				r *= -1
+				c *= -1
+				s *= -1
+			}
+			umat.Data[i*stride+i] = r
+			if i < n-1 {
+				// Multiply the extended factorization matrix by
+				// the Givens matrix from the left. Only
+				// the i-th row and x are modified.
+				blas64.Rot(n-i-1,
+					blas64.Vector{1, umat.Data[i*stride+i+1 : i*stride+n]},
+					blas64.Vector{1, work[i+1 : n]},
+					c, s)
+			}
+		}
+		c.updateCond(-1)
+		return true
+	}
+
+	// Compute rank-1 downdate.
+	alpha = math.Sqrt(-alpha)
+	if alpha != 1 {
+		blas64.Scal(n, alpha, blas64.Vector{1, work})
+	}
+	// Solve U^T * p = x storing the result into work.
+	ok = lapack64.Trtrs(blas.Trans, c.chol.RawTriangular(), blas64.General{
+		Rows:   n,
+		Cols:   1,
+		Stride: 1,
+		Data:   work,
+	})
+	if !ok {
+		// The original matrix is singular.
+		return false
+	}
+	norm := blas64.Nrm2(n, blas64.Vector{1, work})
+	if norm >= 1 {
+		// The updated matrix is not positive definite.
+		return false
+	}
+	norm = math.Sqrt((1 + norm) * (1 - norm))
+	cos := make([]float64, n)
+	sin := make([]float64, n)
+	for i := n - 1; i >= 0; i-- {
+		// Compute parameters of Givens matrices that zero elements of p
+		// backwards.
+		cos[i], sin[i], norm, _ = blas64.Rotg(norm, work[i])
+		if norm < 0 {
+			norm *= -1
+			cos[i] *= -1
+			sin[i] *= -1
+		}
+	}
+	umat := c.chol.mat
+	stride := umat.Stride
+	ok = true
+	for i := n - 1; i >= 0; i-- {
+		// Apply Givens matrices to U.
+		blas64.Rot(n-i, blas64.Vector{1, work[i:n]}, blas64.Vector{1, umat.Data[i*stride+i : i*stride+n]}, cos[i], sin[i])
+		if umat.Data[i*stride+i] == 0 {
```

Comment:
I think not. The decisive check for whether the updated matrix is PD or not is done at line 293. What happens here is that cos and sin are determined so as to zero out the elements of p (work), the elements of U play no role in it. Therefore, when we apply them here to U, the particular combination of sin and cos can produce a negative diagonal element even if the updated matrix is PD. In other words, what this loop does is instead of computing the Cholesky factorization `U^T U = A` it computes a factorization `(DU)^T DU = A` where D is a diagonal matrix with some of its diagonal elements negative. What line 325 does is to undo the matrix D.

Of course, I think it may happen that the updated A is close to singular and that line 293 says that it is PD even if its not, but such case is already covered in the docs by "ill-conditioned" and "inaccurate".